File: rust/src/code/remove_nth_node_from_end_of_list.rs
use super::models::list_node::ListNode;

pub struct Solution {}

impl Solution {
    pub fn remove_nth_from_end(head: Option<Box<ListNode>>, mut n: i32) -> Option<Box<ListNode>> {
        let mut dummy = Box::new(ListNode {
            val: 0,
            next: head.clone(),
        });

        let mut fast = dummy.clone();
        let mut slow = dummy.as_mut();

        while let Some(node) = fast.next.take() {
            if n <= 0 {
                slow = slow.next.as_mut().unwrap();
            }
            fast = node.clone();
            n -= 1;
        }

        slow.next = slow.next.as_mut().unwrap().next.clone();
        return dummy.next;
    }
}

/*
* Note:
* it will move Box<self>, which is allocated on heap, and Rust doesn't allow you to just move values away from heap unless you implement Copy, otherwise what's left there?
* The book uses Option::take() to move ownership out and leave None on the place.
*/

#[cfg(test)]
mod test {
    use super::super::utils::convert_to_linked_list::convert_vec_to_linked_list;
    use super::Solution;

    #[test]
    fn case_1() {
        let head = convert_vec_to_linked_list(vec![1, 2, 3, 4, 5]);
        let n = 2;
        let expected = convert_vec_to_linked_list(vec![1, 2, 3, 5]);
        let actual = Solution::remove_nth_from_end(head, n);

        assert_eq!(actual, expected);

        let head2 = convert_vec_to_linked_list(vec![1]);
        let n2 = 1;
        let expected2 = convert_vec_to_linked_list(vec![]);
        let actual2 = Solution::remove_nth_from_end(head2, n2);

        assert_eq!(actual2, expected2);

        let head3 = convert_vec_to_linked_list(vec![1, 2]);
        let n3 = 1;
        let expected3 = convert_vec_to_linked_list(vec![1]);
        let actual3 = Solution::remove_nth_from_end(head3, n3);

        assert_eq!(actual3, expected3);
    }
}

File: python/remove_nth_node_from_end_of_list.py
from typing import Optional
import sys

sys.path.append("./models")
from list_node import ListNode

sys.path.append("./utils")
from linkedListConverter import convert_to_linked_list
from linkedListConverter import convert_linked_list_to_list


class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        """time complexity: O(n)"""
        dummy: Optional[ListNode] = ListNode()
        dummy.next = head
        slow: Optional[ListNode] = dummy
        fast: Optional[ListNode] = head
        for _ in range(n):
            if fast is None:
                break
            fast = fast.next
        while slow is not None and fast is not None:
            slow = slow.next
            fast = fast.next
        if slow is not None and slow.next is not None:
            slow.next = slow.next.next
        return dummy.next


def test_removeNthFromEnd_case_1():
    # arrange
    head: list[int] = [1, 2, 3, 4, 5]
    n: int = 2
    expected: list[int] = [1, 2, 3, 5]

    # act
    solution = Solution()
    actual = solution.removeNthFromEnd(convert_to_linked_list(head), n)

    # assert
    assert convert_linked_list_to_list(actual) == expected


def test_removeNthFromEnd_case_2():
    # arrange
    head: list[int] = [1]
    n: int = 1
    expected: list[int] = []

    # act
    solution = Solution()
    actual = solution.removeNthFromEnd(convert_to_linked_list(head), n)

    # assert
    assert convert_linked_list_to_list(actual) == expected


def test_removeNthFromEnd_case_3():
    # arrange
    head: list[int] = [1, 2]
    n: int = 1
    expected: list[int] = [1]

    # act
    solution = Solution()
    actual = solution.removeNthFromEnd(convert_to_linked_list(head), n)

    # assert
    assert convert_linked_list_to_list(actual) == expected
